types/json_ml/index.d.ts

```diff
@@ -0,0 +1,20 @@
+// Type definitions for json_ml 0.0
+// Project: https://github.com/59naga/json-ml
+// Definitions by: Alan Plum <https://github.com/pluma>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="cheerio" />
+
+export type Node = string | Element;
```

Comment:
That's a very complicated type but I'm not sure I see how much value it provides.  It won't help with completion since there are no named properties - it looks like it basically just prevents consumers from mixing non-strings into their tree?

Reply:
I wasn't sure how to type this correctly.

Basically the return value from parse is an array of nodes, where each node can be either an element or a string. An element is an array consisting of a tag name as string, optionally an attributes object and then zero or more child nodes.

A naive type would be:

```
type Node = string | [string, Attributes, ...Node[]] | [string, ...Node[]];
function parse(html: string): Node[];
```

but this is self-referential so I had to use an interface.

Sadly the attributes object is only present when the element has attributes, otherwise this would be a lot easier to handle.

In my own code I'm consuming node lists recursively, so for every node I have the following checks:

* `typeof node === "string"`
* `node[1] && typeof node[1] === "object" && !Array.isArray(node[1])` (to check if the element has attributes)

The typings help a bit IMO.

Reply:
Hm... looks like the typings don't actually work correctly.

Reply:
Thanks, I think the simple version is a good balance between clarity and strictness.